smarts/core/opendrive_road_network.py

```diff
@@ -0,0 +1,1812 @@
+
+                if lane_elem.id < 0:
+                    # Direction of lane is the same as the reference line
+                    if pred_lane not in lane.incoming_lanes:
+                        lane.incoming_lanes.append(pred_lane)
+                    if lane not in pred_lane.outgoing_lanes:
+                        pred_lane.outgoing_lanes.append(lane)
+                else:
+                    # Direction of lane is opposite the refline, so this is actually an outgoing lane
+                    if pred_lane not in lane.outgoing_lanes:
+                        lane.outgoing_lanes.append(pred_lane)
+                    if lane not in pred_lane.incoming_lanes:
+                        pred_lane.incoming_lanes.append(lane)
+
+        if lane_link.successorId:
+            road_id, section_id = None, None
+            if ls_index == len(road_elem.lanes.lane_sections) - 1:
+                # This is the last lane section, so get the first/last lane section of the successor road
+                road_successor = road_elem.link.successor
+                if road_successor and road_successor.elementType == "road":
+                    road_id = road_successor.element_id
+                    succ_road_elem = od.getRoad(road_id)
+                    if not succ_road_elem:
+                        self._log.warning(
+                            f"Successor road {road_id} does not exist for road {road_elem.id}"
+                        )
+                        return
+                    section_id = (
+                        succ_road_elem.lanes.getLastLaneSectionIdx()
+                        if road_successor.contactPoint == "end"
+                        else 0
+                    )
+            else:
+                # Otherwise, get the next lane section in the current road
+                road_id = road_elem.id
+                section_id = ls_index + 1
+
+            if road_id is not None and section_id is not None:
+                succ_suffix = "L" if lane_link.successorId > 0 else "R"
+                succ_lane_id = (
+                    f"{road_id}_{section_id}_{succ_suffix}_{lane_link.successorId}"
+                )
+                succ_lane = self.lane_by_id(succ_lane_id)
+
+                if lane_elem.id < 0:
+                    # Direction of lane is the same as the reference line
+                    if succ_lane not in lane.outgoing_lanes:
+                        lane.outgoing_lanes.append(succ_lane)
+                    if lane not in succ_lane.incoming_lanes:
+                        succ_lane.incoming_lanes.append(lane)
+                else:
+                    # Direction of lane is opposite the refline, so this is actually an incoming lane
+                    if succ_lane not in lane.incoming_lanes:
+                        lane.incoming_lanes.append(succ_lane)
+                    if lane not in succ_lane.outgoing_lanes:
+                        succ_lane.outgoing_lanes.append(lane)
```

Comment:
They may appear to be duplicate but both `successorId` vs `predecessorId` blocks are very different. The subsequent nested conditions are also different, and the statements we execute and parameters and variables we set are also different. 

I am not sure how this may seem duplicate. If we try to condense them into one block, we will end up checking `successorId` vs `predecessorId`, and `road_elem.link.successor` vs `road_elem.link.predecessor` `ls_index == len(road_elem.lanes.lane_sections) - 1` vs `ls_index == 0`  and so on multiple times leading to more if and else statements.

This was our final iteration after refactoring this method 5 to 6 times alone.

Reply:
I see, yes, it mostly took me that the shape of the two blocks are about identical (despite the conditions being different.) I was considering if the shape could be described by local methods expressing intention.

Reply:
We can try breaking into methods, but they wont be used by any other method leading to unnecessary method clutter.